File: examples/standard_devices/robot/cs/robot_client_jog_freespace.cs
using System;
using RobotRaconteur;
using System.Threading;
using System.Threading.Tasks;
using System.Linq;
using com.robotraconteur.robotics.robot;
using com.robotraconteur.robotics.trajectory;
using System.Diagnostics;

// URL for connecting to the robot. By default connects to UR5e simulated robot.
string url = "rr+tcp://localhost:52511?service=robot";
if (args.Length >= 2)
{
    url = args[1];
}

// Initialize the client node
using (var node_setup = new ClientNodeSetup())
{
    // Connect to the robot
    var c = (Robot)RobotRaconteurNode.s.ConnectService(url);

    // Retrieve the current robot state and print the current command mode
    var robot_state = c.robot_state.PeekInValue(out var state_ts);
    Console.WriteLine(robot_state.command_mode);

    // Change the robot command mode, first to halt, then to jog
    c.command_mode = RobotCommandMode.halt;
    Thread.Sleep(100);
    c.command_mode = RobotCommandMode.jog;

    // Get the starting joint positions
    var robot_state2 = c.robot_state.PeekInValue(out var state_ts2);
    var start_joint_pos = robot_state2.joint_position;

    var sw = new Stopwatch();
    sw.Start();

    TimeSpan t_start = sw.Elapsed;

    // Move the robot in a loop
    for (int i = 0; i < 100; i++)
    {
        var t = (double)((sw.Elapsed - t_start).TotalMilliseconds) / 1000.0;

        // Generate a sin wave
        var cmd = new double[6];
        for (int j = 0; j < 6; j++)
        {
            cmd[j] = start_joint_pos[j];
        }
        cmd[0] += 0.1 * Math.Sin(t / 5.0);

        // Send the command to the robot
        c.jog_freespace(cmd, new double[6] { 1.0, 1.0, 1.0, 1.0, 1.0, 1.0 }, true);

        // Wait for the next loop iteration
        Thread.Sleep(100);
    }
}

File: examples/standard_devices/robot/python/robot_client_jog_freespace.py
##
# Command the robot using the jog_freespace function
#

from RobotRaconteur.Client import *
import time
import numpy as np
import sys

# URL for connecting to the robot. By default connects to UR5e simulated robot.
url = 'rr+tcp://localhost:52511?service=robot'
if (len(sys.argv) >= 2):
    url = sys.argv[1]

# Connect to the robot driver
c = RRN.ConnectService(url)

# Retrieve the current robot state and print the current command mode
print(c.robot_state.PeekInValue()[0].command_mode)

# Retrieve the constants for the com.robotraconteur.robotics.robot service definition
robot_const = RRN.GetConstants("com.robotraconteur.robotics.robot", c)

# Retrieve the "halt" and "jog" enum values
halt_mode = robot_const["RobotCommandMode"]["halt"]
jog_mode = robot_const["RobotCommandMode"]["jog"]

# Change the robot command mode, first to halt, then to jog
c.command_mode = halt_mode
time.sleep(0.1)
c.command_mode = jog_mode


# Get the starting joint positions
robot_state = c.robot_state.PeekInValue()[0]
start_joint_pos = robot_state.joint_position

t_start = time.time()

# Move the robot in an endless loop
while (True):
    t = time.time() - t_start

    # Generate a sin wave
    c.jog_freespace(0.2 * np.array([1, 0, 0, 0, 0, 0]) * np.sin(t / 5) + start_joint_pos, np.ones((6,)), True)

    # Print the current robot_state_flags
    print(hex(c.robot_state.PeekInValue()[0].robot_state_flags))

    # Wait for the next loop iteration
    time.sleep(0.1)
